mat64/shadow.go

```diff
@@ -0,0 +1,201 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat64
+
+import "github.com/gonum/blas/blas64"
+
+const (
+	// regionOverlap is the panic string used for the general case
+	// of a matrix region overlap between a source and destination.
+	regionOverlap = "mat64: bad region: overlap"
+
+	// regionIdentity is the panic string used for the specific
+	// case of complete agreement between a source and a destination.
+	regionIdentity = "mat64: bad region: identical"
+
+	// mismatchedStrides is the panic string used for overlapping
+	// data slices with differing strides.
+	mismatchedStrides = "mat64: bad region: different strides"
+)
+
+// checkOverlap returns false if the receiver does not overlap data elements
+// referenced by the parameter and panics otherwise.
+//
+// checkOverlap methods return a boolean to allow the check call to be added to a
+// boolean expression, making use of short-circuit operators.
+
+func (m *Dense) checkOverlap(a blas64.General) bool {
```

Comment:
It's easy to test for all Raw types, isn't it? If this takes in Matrix, then a type switch can be used to extract the data slice from any blas type. If there is any overlap and the types are different, we panic. If both are RawMatrixers, then we can do the extended testing.